File: BackEnd/Friend Finder System/src/main/java/ahmed/com/springboot/friend_finder_system/service/Comment_Service.java
package ahmed.com.springboot.friend_finder_system.service;

import ahmed.com.springboot.friend_finder_system.Vm.CommentRequest_Vm;
import ahmed.com.springboot.friend_finder_system.Vm.UpdateCommentRequest_Vm;
import ahmed.com.springboot.friend_finder_system.dto.CommentDto;

import java.util.List;

public interface Comment_Service {

    Void createComment(CommentRequest_Vm commentRequestVm);

    void deleteComment(Long commentId);

    void updateComment(UpdateCommentRequest_Vm CommentRequestVm);

    void replyToComment(UpdateCommentRequest_Vm commentRequestVm);

    List<CommentDto> getComments(Long postId, int pageNumber);

    List<CommentDto> getReplies(Long commentId , int pageNumber);

}

File: frontEnd/friendFinder/src/app/core/services/comments/comment-service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface CommentAuthor {
  id: number;
  firstName: string;
  lastName: string;
  profilePicture: string | null;
  city: string;
}

export interface CommentDto {
  id: number;
  content: string;
  author: CommentAuthor;
  postId: number;
  likedIs: number | null;
  countComments: number | null;
  localDateTime: string;
}

export interface NewCommentRequest {
  postId: number;
  content: string;
}

export interface UpdateCommentRequest {
  commentId: number;
  content: string;
}

export interface ReplyRequest {
  commentId: number;
  content: string;
}

@Injectable({
  providedIn: 'root'
})
export class CommentService {

  private http = inject(HttpClient);
  private readonly API_URL = 'http://localhost:8081/api/comments';

  getCommentsByPostId(postId: number, pageNumber: number = 1): Observable<CommentDto[]> {
    const params = new HttpParams().set('pageNumber', pageNumber.toString());
    return this.http.get<CommentDto[]>(
      `${this.API_URL}/CommentsByPostId/${postId}`,
      { params }
    );
  }

  addComment(request: NewCommentRequest): Observable<any> {
    return this.http.post(`${this.API_URL}/newComment`, request);
  }

  updateComment(request: UpdateCommentRequest): Observable<void> {
    return this.http.put<void>(`${this.API_URL}/updateComment`, request);
  }

  replyToComment(request: ReplyRequest): Observable<void> {
    return this.http.post<void>(`${this.API_URL}/replyToComment`, request);
  }

  deleteComment(commentId: number): Observable<void> {
    return this.http.delete<void>(`${this.API_URL}/deleteComment/${commentId}`);
  }

  getReplies(commentId: number, pageNumber: number = 1): Observable<CommentDto[]> {
    const params = new HttpParams().set('pageNumber', pageNumber.toString());
    return this.http.get<CommentDto[]>(
      `${this.API_URL}/replies/${commentId}`,
      { params }
    );
  }
}
